CDAP-3750 escape hive column names

albertshau opened the pull request:
Some common column names, like 'user', are reserved Hive keywords.
Depending on the version of Hive used and various configuration
settings, ddl statements need to have their columns escaped with
backticks in case one of the reserved keywords is used.

albertshau committed:
CDAP-3750 escape hive column names

Some common column names, like 'user', are reserved Hive keywords.
Depending on the version of Hive used and various configuration
settings, ddl statements need to have their columns escaped with
backticks in case one of the reserved keywords is used.

albertshau commented:
http://builds.cask.co/browse/CDAP-DUT4122

poornachandra commented:
One concern with quoting columns is that when the table is described, it is not clear that the column needs to be quoted in queries. This can confuse users when queries fail due to parsing error. There is no good solution for this. Maybe just document this behavior then?

See discussion in https://issues.cask.co/browse/CDAP-5629

albertshau commented:
I understand the reasoning for changing the column names for things under the control of the system, but its even more confusing for a user if they deploy their dataset and it doesn't show up as explorable.  At least for the unescaped query, you'll get the Hive message back, and its somewhat clear that is some identifier error.

What we can do though, is change the 'date' field in the examples to something else so that people don't run into it for those examples.  As well as documenting this behavior like you suggested.

poornachandra commented:
+1

awholegunch commented:
In a recent PR (https://github.com/caskdata/cdap/pull/5754 https://github.com/caskdata/cdap/pull/5754) I added a note and a link to the SportsResults example pointing out the usage of back-ticks for passing Hive reserved keywords.

> On May 9, 2016, at 12:13 PM, poornachandra notifications@github.com wrote:
> 
> +1
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub https://github.com/caskdata/cdap/pull/5752#issuecomment-217960556

poornachandra commented on cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java:
```diff
@@ -461,7 +468,13 @@ private String generateHivePartitionKey(PartitionKey key) {
-      builder.append(sep).append(fieldName).append("=").append(quote).append(fieldValue.toString()).append(quote);
+      builder.append(sep);
+      if (shouldEscapeColumns) {
+        builder.append('`').append(fieldName).append('`');
```
Will this quote all the columns?

anew replied:
I think it will but that should not do a harm.

anew replied:
There is one tricky issue here, if a column name contains a ' character which apparently was allowed prior to Hive 12. That should never happen in CDAP because as far as I know we do not allow ' in field names of a schema.

albertshau replied:
I can just check for it and escape accordingly. According to the docs, a literal backtick is just two backticks.

anew commented on cdap-explore/src/test/java/co/cask/cdap/explore/table/SchemaConverterTest.java:
```diff
@@ -210,13 +225,13 @@ public void testUnsupportedTypes() {
   public void testSupportedTypes() throws Exception {
     // Should not throw an exception
-    SchemaConverter.toHiveSchema(NotRecursive.class);
+    new SchemaConverter(false).toHiveSchema(NotRecursive.class);
```
I think this only tests the case where escaping is on. Should this also test the other case?

anew commented on cdap-explore/src/test/java/co/cask/cdap/explore/table/SchemaConverterTest.java:
```diff
@@ -210,13 +225,13 @@ public void testUnsupportedTypes() {
     String schema;
     try {
-      schema = SchemaConverter.toHiveSchema(type);
+      schema = new SchemaConverter(false).toHiveSchema(type);
```
same here

anew commented:
The consensus appears to be that it is better to change all examples not to use reserved words. I am not sure that is the best thing... it may make more sense to have examples use reserved words _and_ explain in the Readme/docs that this field must be escaped in queries. Otherwise, users will not be aware of it at all, and when they create their own datasets, they will use reserved words, and be even more confused.

awholegunch commented:
+1 for showing how to use reserved keywords properly.

albertshau commented:
That's a good point, I will keep the sport example with 'date' as is.  JJ has already added a note in the docs about reserved keywords for that example.

But if there are any CDAP system datasets, stream formats, or hydrator plugins that use a hardcoded reserved keyword, we should see if we should rename those fields.  The only one I know of is the DataGenerator realtime source, that uses 'binary' as a field name. 

In addition, I will add a note about reserved keywords on the doc about custom explorable datasets.

albertshau committed:
CDAP-3750 escape backticks in hive column names and add hive-site

Escaping backticks within a column name.
Also adding hive-site.xml to the hadoop conf for the explore master
service so that the hConf instance actually includes the settings
from hive-site.xml.

Also adding docs about enclosing reserved keywords in backticks.

anew commented:
LGTM
